File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RankingsLayout.java
package com.garpr.android.views;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.adapters.RankingsAdapter;
import com.garpr.android.misc.ListUtils;
import com.garpr.android.misc.MiscUtils;
import com.garpr.android.misc.Refreshable;
import com.garpr.android.misc.RegionManager;
import com.garpr.android.misc.ThreadUtils;
import com.garpr.android.models.Ranking;
import com.garpr.android.models.RankingsBundle;
import com.garpr.android.networking.ApiCall;
import com.garpr.android.networking.ApiListener;
import com.garpr.android.networking.ServerApi;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;

public class RankingsLayout extends SearchableFrameLayout implements ApiListener<RankingsBundle>,
        Refreshable, SwipeRefreshLayout.OnRefreshListener {

    private RankingsAdapter mAdapter;
    private RankingsBundle mRankingsBundle;

    @Inject
    RegionManager mRegionManager;

    @Inject
    ServerApi mServerApi;

    @Inject
    ThreadUtils mThreadUtils;

    @BindView(R.id.recyclerView)
    RecyclerView mRecyclerView;

    @BindView(R.id.refreshLayout)
    SwipeRefreshLayout mRefreshLayout;

    @BindView(R.id.empty)
    View mEmpty;

    @BindView(R.id.error)
    View mError;


    public static RankingsLayout inflate(final ViewGroup parent) {
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        return (RankingsLayout) inflater.inflate(R.layout.layout_rankings, parent, false);
    }

    public RankingsLayout(@NonNull final Context context, @Nullable final AttributeSet attrs) {
        super(context, attrs);
    }

    public RankingsLayout(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public RankingsLayout(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr, @StyleRes final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public void failure(final int errorCode) {
        mRankingsBundle = null;
        onRankingsBundleFetched();
        showError();
    }

    private void fetchRankingsBundle() {
        mRankingsBundle = null;
        mRefreshLayout.setRefreshing(true);
        mServerApi.getRankings(mRegionManager.getRegion(getContext()), new ApiCall<>(this));
    }

    @Nullable
    public RankingsBundle getRankingsBundle() {
        return mRankingsBundle;
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        if (isInEditMode()) {
            return;
        }

        App.get().getAppComponent().inject(this);

        mRefreshLayout.setOnRefreshListener(this);
        mRecyclerView.addItemDecoration(new DividerItemDecoration(getContext(),
                DividerItemDecoration.VERTICAL));
        mRecyclerView.setHasFixedSize(true);
        mAdapter = new RankingsAdapter(getContext());
        mRecyclerView.setAdapter(mAdapter);

        fetchRankingsBundle();
    }

    private void onRankingsBundleFetched() {
        if (!isAlive()) {
            return;
        }

        final Activity activity = MiscUtils.optActivity(getContext());

        if (activity instanceof Listener) {
            ((Listener) activity).onRankingsBundleFetched(this);
        }
    }

    @Override
    public void onRefresh() {
        fetchRankingsBundle();
    }

    @Override
    public void refresh() {
        fetchRankingsBundle();
    }

    @Override
    public void search(@Nullable final String query) {
        mThreadUtils.run(new ThreadUtils.Task() {
            private List<Ranking> mList;

            @Override
            public void onBackground() {
                if (!isAlive() || !TextUtils.equals(query, getSearchQuery())) {
                    return;
                }

                mList = ListUtils.searchRankingList(query, mRankingsBundle.getRankings());
            }

            @Override
            public void onUi() {
                if (!isAlive() || !TextUtils.equals(query, getSearchQuery())) {
                    return;
                }

                mAdapter.set(mList);
            }
        });
    }

    private void showEmpty() {
        mAdapter.clear();
        mRecyclerView.setVisibility(View.GONE);
        mError.setVisibility(View.GONE);
        mEmpty.setVisibility(View.VISIBLE);
        mRefreshLayout.setRefreshing(false);
    }

    private void showError() {
        mAdapter.clear();
        mRecyclerView.setVisibility(View.GONE);
        mEmpty.setVisibility(View.GONE);
        mError.setVisibility(View.VISIBLE);
        mRefreshLayout.setRefreshing(false);
    }

    private void showRankingsBundle() {
        mAdapter.set(mRankingsBundle);
        mEmpty.setVisibility(View.GONE);
        mError.setVisibility(View.GONE);
        mRecyclerView.setVisibility(View.VISIBLE);
        mRefreshLayout.setRefreshing(false);
    }

    @Override
    public void success(@Nullable final RankingsBundle rankingsBundle) {
        mRankingsBundle = rankingsBundle;
        onRankingsBundleFetched();

        if (mRankingsBundle != null && mRankingsBundle.hasRankings()) {
            showRankingsBundle();
        } else {
            showEmpty();
        }
    }


    public interface Listener {
        void onRankingsBundleFetched(final RankingsLayout layout);
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RankingsLayout.kt
package com.garpr.android.views

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.garpr.android.R
import com.garpr.android.adapters.RankingsAdapter
import com.garpr.android.extensions.activity
import com.garpr.android.extensions.appComponent
import com.garpr.android.managers.NotificationsManager
import com.garpr.android.managers.RegionManager
import com.garpr.android.misc.ListUtils
import com.garpr.android.misc.Refreshable
import com.garpr.android.misc.ThreadUtils
import com.garpr.android.models.RankedPlayer
import com.garpr.android.models.RankingsBundle
import com.garpr.android.networking.ApiCall
import com.garpr.android.networking.ApiListener
import com.garpr.android.networking.ServerApi
import kotlinx.android.synthetic.main.layout_rankings.view.*
import javax.inject.Inject

class RankingsLayout @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : SearchableRefreshLayout(context, attrs), ApiListener<RankingsBundle>, Refreshable,
        SwipeRefreshLayout.OnRefreshListener {

    @Inject
    protected lateinit var notificationsManager: NotificationsManager

    @Inject
    protected lateinit var regionManager: RegionManager

    @Inject
    protected lateinit var serverApi: ServerApi

    private lateinit var adapter: RankingsAdapter

    var rankingsBundle: RankingsBundle? = null
        private set


    interface Listener {
        fun onRankingsBundleFetched(layout: RankingsLayout)
    }

    companion object {
        fun inflate(parent: ViewGroup): RankingsLayout = LayoutInflater.from(parent.context)
                .inflate(R.layout.layout_rankings, parent, false) as RankingsLayout
    }

    override fun failure(errorCode: Int) {
        rankingsBundle = null
        notificationsManager.cancelAll()
        onRankingsBundleFetched()
        showError(errorCode)
    }

    private fun fetchRankingsBundle() {
        isRefreshing = true
        serverApi.getRankings(regionManager.getRegion(context), ApiCall(this))
    }

    override fun getRecyclerView(): RecyclerView? {
        return recyclerView
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        if (isInEditMode) {
            return
        }

        appComponent.inject(this)

        setOnRefreshListener(this)
        recyclerView.addItemDecoration(DividerItemDecoration(context,
                DividerItemDecoration.VERTICAL))
        recyclerView.setHasFixedSize(true)
        adapter = RankingsAdapter(context)
        recyclerView.adapter = adapter

        fetchRankingsBundle()
    }

    private fun onRankingsBundleFetched() {
        if (isAlive) {
            (activity as? Listener)?.onRankingsBundleFetched(this)
        }
    }

    override fun onRefresh() {
        fetchRankingsBundle()
    }

    override fun refresh() {
        notificationsManager.cancelAll()
        fetchRankingsBundle()
    }

    override fun search(query: String?) {
        val rankings = rankingsBundle?.rankings

        if (rankings.isNullOrEmpty()) {
            return
        }

        threadUtils.run(object : ThreadUtils.Task {
            private var list: List<RankedPlayer>? = null

            override fun onBackground() {
                if (!isAlive || !TextUtils.equals(query, searchQuery)) {
                    return
                }

                list = ListUtils.searchRankingList(query, rankings)
            }

            override fun onUi() {
                if (!isAlive || !TextUtils.equals(query, searchQuery)) {
                    return
                }

                adapter.set(list)
            }
        })
    }

    private fun showEmpty() {
        adapter.clear()
        recyclerView.visibility = View.GONE
        error.visibility = View.GONE
        empty.visibility = View.VISIBLE
        isRefreshing = false
    }

    private fun showError(errorCode: Int) {
        adapter.clear()
        recyclerView.visibility = View.GONE
        empty.visibility = View.GONE
        error.setVisibility(View.VISIBLE, errorCode)
        isRefreshing = false
    }

    private fun showRankingsBundle() {
        adapter.set(rankingsBundle)
        empty.visibility = View.GONE
        error.visibility = View.GONE
        recyclerView.visibility = View.VISIBLE
        isRefreshing = false
    }

    override fun success(`object`: RankingsBundle?) {
        rankingsBundle = `object`
        notificationsManager.cancelAll()

        if (`object`?.rankings?.isNotEmpty() == true) {
            showRankingsBundle()
        } else {
            showEmpty()
        }

        onRankingsBundleFetched()
    }

}
